<img src="https://raw.githubusercontent.com/andre-marcos-perez/ebac-course-utils/main/media/logo/newebac_logo_black_half.png" alt="ebac-logo">

---

# **Módulo** | Análise de Dados: Fundamentos de Estatística
Caderno de **Exercícios**<br>
Professor [André Perez](https://www.linkedin.com/in/andremarcosperez/)

---

# **Tópicos**

<ol type="1">
  <li>Média e Variância;</li>
  <li>Ordem e Posição;</li>
  <li>Correlação.</li>
</ol>

---

# **Exercícios**

## 1\. Tráfego de São Paulo

Neste exercício, vamos continuar a analisar os dados de mobilidade urbana da cidade de São Paulo. A base de dados está neste [link](https://raw.githubusercontent.com/andre-marcos-perez/ebac-course-utils/develop/dataset/traffic.csv) e é uma copia do dado original, presente neste [link](https://archive.ics.uci.edu/ml/datasets/Behavior+of+the+urban+traffic+of+the+city+of+Sao+Paulo+in+Brazil). A base de dados contem a quantidade de acidentes ocorridos na cidade entre 14/12/09 e 18/12/09, das 07:00h ás 20:00h, agregados em intervalos de 30 minutos.

In [2]:
!wget -q 'https://raw.githubusercontent.com/andre-marcos-perez/ebac-course-utils/develop/dataset/traffic.csv' -O traffic.csv

In [3]:
import pandas as pd

In [4]:
df = pd.read_csv('traffic.csv', sep=';')

In [5]:
df.head()

hour  bus_stoped  broken_truck  vehicle_excess  accident_victim  \
0     1           0             0               0                0   
1     2           0             0               0                0   
2     3           0             0               0                0   
3     4           0             0               0                0   
4     5           0             0               0                0   

   running_over  fire_vehicles  occurrence_freight  \
0             0              0                   0   
1             0              0                   0   
2             0              0                   0   
3             0              0                   0   
4             0              0                   0   

   incident__dangerous_freight  lack_electricity  fire  flooding  \
0                            0                 0     0         0   
1                            0                 0     0         0   
2                            0                 0     0         0   
3                            0                 0     0         0   
4                            0                 0     0         0   

   manifestations  defect_trolleybuses  tree_road  semaphore_off  \
0               0                    0          0              0   
1               0                    0          0              0   
2               0                    0          0              0   
3               0                    0          0              0   
4               0                    0          0              0   

   intermittent_emaphore slowness_traffic_%  
0                      0                4,1  
1                      0                6,6  
2                      0                8,7  
3                      0                9,2  
4                      0               11,1

- **Exemplo**: (para relembrar)

O código abaixo extrai a 13ª linha do arquivo que representa a 14ª meia hora do dia 14/12/09, contadas a partir das 07:00h, ou seja, todos os incidentes ocorridos na cidade no dia em questão entre as 13:30h e 14:00h. Vemos, por exemplo, que dois ônibus e dois caminhões ficaram paradados nas ruas da cidade, atrapalhando o tráfego.

In [6]:
df.iloc[[13]]

hour  bus_stoped  broken_truck  vehicle_excess  accident_victim  \
13    14           2             2               0                1   

    running_over  fire_vehicles  occurrence_freight  \
13             0              0                   0   

    incident__dangerous_freight  lack_electricity  fire  flooding  \
13                            0                 0     0         0   

    manifestations  defect_trolleybuses  tree_road  semaphore_off  \
13               0                    0          0              0   

    intermittent_emaphore slowness_traffic_%  
13                      0                4,9

### **1.1. Agregação**

Neste primeira atividade, você deve gerar um array NumPy por dia. Para cada array você deve somar todos os incidentes que aconteceram naquela meia hora. Sendo assim, cada array deve ter 27 posições, cada qual com a soma dos incidentes daquela meira hora.

**Dica**: Você deve remover a primeira e a última coluna.

**Dica**: Os arrays devem ter os seguintes valores:

```python
14: [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 5, 2, 0, 0, 0, 0, 0, 1, 0, 0, 0, 2, 2, 3]
15: [3, 0, 0, 0, 2, 4, 2, 1, 3, 0, 2, 0, 2, 3, 2, 2, 4, 7, 6, 3, 0, 1, 1, 5, 4, 6, 0]
16: [1, 1, 2, 4, 4, 5, 3, 2, 4, 2, 1, 1, 3, 6, 2, 0, 2, 8, 4, 6, 4, 1, 2, 7, 15, 4, 0]
17: [4, 1, 1, 0, 0, 0, 2, 3, 2, 7, 2, 1, 7, 3, 6, 4, 3, 2, 2, 3, 2, 5, 2, 0, 1, 4, 15]
18: [10, 1, 1, 5, 5, 7, 3, 4, 3, 1, 2, 3, 2, 2, 6, 2, 2, 4, 2, 3, 0, 3, 3, 2, 3, 6, 0]
```

In [59]:
# resposta da questão 1.1

import numpy as np
import pandas as pd

df = pd.read_csv('traffic.csv', encoding='utf8', sep=';')

print(df.head()) #exibe as primeiras linhas

   hour  bus_stoped  broken_truck  vehicle_excess  accident_victim  \
0     1           0             0               0                0   
1     2           0             0               0                0   
2     3           0             0               0                0   
3     4           0             0               0                0   
4     5           0             0               0                0   

   running_over  fire_vehicles  occurrence_freight  \
0             0              0                   0   
1             0              0                   0   
2             0              0                   0   
3             0              0                   0   
4             0              0                   0   

   incident__dangerous_freight  lack_electricity  fire  flooding  \
0                            0                 0     0         0   
1                            0                 0     0         0   
2                            0                 0  

In [60]:
print(df.shape) #exibe o formato/quantidade (linhas, colunas)
print(df.columns) #exibe o nome das colunas

(135, 18)
Index(['hour', 'bus_stoped', 'broken_truck', 'vehicle_excess',
       'accident_victim', 'running_over', 'fire_vehicles',
       'occurrence_freight', 'incident__dangerous_freight', 'lack_electricity',
       'fire', 'flooding', 'manifestations', 'defect_trolleybuses',
       'tree_road', 'semaphore_off', 'intermittent_emaphore',
       'slowness_traffic_%'],
      dtype='object')


In [61]:
# Remove colunas desnecessárias (a primeira 'hour' e a última 'slowness_traffic_%')
df = df.iloc[:, 1:-1]

# Converte os dados para valores numéricos (substituir vírgula por ponto se necessário)
df = df.map(lambda x: str(x).replace(",", "."))
df = df.astype(float) #para corrigir possíveis vírgulas em números decimais

# Determina o número de dias no conjunto de dados
n_intervalos = 27  #considerando que há 27 intervalos de 30 minutos por dia
n_dias = len(df) // n_intervalos  #número de dias

# Cria os arrays diários
arrays_diarios = []

for i in range(n_dias):
    inicio = i * n_intervalos
    fim = inicio + n_intervalos
    soma_incidentes = df.iloc[inicio:fim, :].sum(axis=1).to_numpy()
    arrays_diarios.append(soma_incidentes)

# Exibe a saída
for i, array in enumerate(arrays_diarios):
    print(f'Dia {i+1}: {array}')

Dia 1: [0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 1. 0. 5. 2. 0. 0. 0. 0. 0. 1. 0. 0. 0.
 2. 2. 3.]
Dia 2: [3. 0. 0. 0. 2. 4. 2. 1. 3. 0. 2. 0. 2. 3. 2. 2. 4. 7. 6. 3. 0. 1. 1. 5.
 4. 6. 0.]
Dia 3: [ 1.  1.  2.  4.  4.  5.  3.  2.  4.  2.  1.  1.  3.  6.  2.  0.  2.  8.
  4.  6.  4.  1.  2.  7. 15.  4.  0.]
Dia 4: [ 4.  1.  1.  0.  0.  0.  2.  3.  2.  7.  2.  1.  7.  3.  6.  4.  3.  2.
  2.  3.  2.  5.  2.  0.  1.  4. 15.]
Dia 5: [10.  1.  1.  5.  5.  7.  3.  4.  3.  1.  2.  3.  2.  2.  6.  2.  2.  4.
  2.  3.  0.  3.  3.  2.  3.  6.  0.]


### **1.2. Métricas**

Para cada array você deve calcular as seguintes métricas:

 - **média**;
 - **desvio padrão**.

In [69]:
# resposta da questão 1.2

import numpy as np
import pandas as pd

# Carrega os dados
df = pd.read_csv('traffic.csv', encoding='utf8', sep=';')

# Remove colunas desnecessárias (a primeira 'hour' e a última 'slowness_traffic_%')
df = df.iloc[:, 1:-1]

# Converte os dados para valores numéricos (substituir vírgula por ponto se necessário)
df = df.map(lambda x: str(x).replace(",", "."))
df = df.astype(float) #para corrigir possíveis vírgulas em números decimais

# Determina o número de dias no conjunto de dados
n_intervalos = 27  #considerando que há 27 intervalos de 30 minutos por dia
n_dias = len(df) // n_intervalos  #número de dias

# Cria os arrays diários
arrays_diarios = []

for i in range(n_dias):
    inicio = i * n_intervalos
    fim = inicio + n_intervalos
    soma_incidentes = df.iloc[inicio:fim, :].sum(axis=1).to_numpy()
    arrays_diarios.append(soma_incidentes)

# Calcula as métricas (média e desvio padrão)
metricas = []

for array in arrays_diarios:
  media = np.mean(array) #np.mean calcula a média do array
  desvio_padrao = np.std(array) #np.std calcula o desvio padrão para o array

  metricas.append((media, desvio_padrao)) #armazena as métricas de cada dia

# Exibe as métricas para cada dia
for i, (media, desvio_padrao) in enumerate(metricas):
  print(f'Dia {i+1}: Média = {media: }, Desvio Padrão = {desvio_padrao: }')

Dia 1: Média =  0.5925925925925926, Desvio Padrão =  1.1944085553034889
Dia 2: Média =  2.3333333333333335, Desvio Padrão =  2.0184335693983275
Dia 3: Média =  3.4814814814814814, Desvio Padrão =  3.047407767514698
Dia 4: Média =  3.037037037037037, Desvio Padrão =  3.0487578707046925
Dia 5: Média =  3.1481481481481484, Desvio Padrão =  2.1892617342417258


In [70]:
# Exibe as métricas com duas casas decimais
for i, (media, desvio_padrao) in enumerate(metricas):
  print(f'Dia {i+1}: Média = {media:.2f}, Desvio Padrão = {desvio_padrao:.2f}')

Dia 1: Média = 0.59, Desvio Padrão = 1.19
Dia 2: Média = 2.33, Desvio Padrão = 2.02
Dia 3: Média = 3.48, Desvio Padrão = 3.05
Dia 4: Média = 3.04, Desvio Padrão = 3.05
Dia 5: Média = 3.15, Desvio Padrão = 2.19


### **1.3. Interpretação**

Baseado nos resultados da questão 1.2, responda:

> Qual dia apresenta a maior média de acidentes por meia hora?

**Resposta:**

A maior média de acidentes por meia hora está no Dia 3 com uma média de 3.48.

> Qual dia apresenta a menor variação de acidentes por meia hora?

**Resposta:**

A menor variação de acidentes por meia hora está no Dia 1, com um desvio padrão de 1.19.

---